Update cache guide to including the caching for traced models (#557)
huggingface/optimum-neuron: docs/source/guides/cache_system.mdx

# Neuron Model Cache

The Neuron Model Cache is a remote cache for compiled Neuron models in the `neff` or `.neuron`(TorchScript) format.
It is integrated into the `NeuronTrainer` and `NeuronModelForXXX` classes to enable loading pretrained models from the cache instead of compiling them locally.

The Neuron Model Cache is hosted on the [Hugging Face Hub](https://huggingface.co/aws-neuron/optimum-neuron-cache) and includes compiled files for all popular and supported `optimum-neuron` pre-trained models.

Before training a Transformers or Diffusion model or loading a NeuronModelForXXX on Neuron platforms, it needs to be exported to neuron format
with [`torch-neuronx`](https://github.com/aws-neuron/aws-neuron-samples/tree/master/torch-neuronx).

When exporting a model, [`torch-neuronx`](https://github.com/aws-neuron/aws-neuron-samples/tree/master/torch-neuronx) will either:

> - 1st approach: convert it to a set of [XLA](https://github.com/pytorch/xla/) subgraphs, and compile each subgraph with the neuronx compiler into a Neuron Executable File Format (NEFF) binary file(eg. training and inference of llm). 

Reviewer: ```suggestion
- 1st approach: convert it to a set of [XLA](https://github.com/pytorch/xla/) subgraphs, and compile each subgraph with the neuronx compiler into a Neuron Executable File Format (NEFF) binary file (eg. training and inference of llm). 
```

Or

> - 2nd approach: recorde one graph, compile it with the neuronx-cc compiler to a NEFF binary file, and serialize as a TorchScript module(eg. inference of encoders, stable diffusion and seq2seq models).

Reviewer: ```suggestion
- 2nd approach: record one graph, compile it with the neuronx-cc compiler to a NEFF binary file, and serialize as a TorchScript module (eg. inference of encoders, stable diffusion and seq2seq models).
```

> - 2nd approach: recorde one graph, compile it with the neuronx-cc compiler to a NEFF binary file, and serialize as a TorchScript module(eg. inference of encoders, stable diffusion and seq2seq models).

Reviewer 2: ```suggestion
- 2nd approach: recorde one graph via tracing (TorchScript), compile it with the neuronx compiler to a NEFF binary file, and serialize as a TorchScript module(eg. inference of encoders, stable diffusion and seq2seq models).
```
Between the two bullet points you use "neuronx compiler" and "neuronx-cc compiler". I think the former is more accurate.

The graph recording is relatively fast, but the compilation takes a lot of time.

To avoid recompiling all NEFF files, for models adopting the first approach, every time a model is loaded on a NeuronX host, [`torch-neuronx`](https://github.com/aws-neuron/aws-neuron-samples/tree/master/torch-neuronx)

>  stores NEFF files in a local directory, usually `/var/tmp/neuron-compile-cache`. And for models adopting the second approach, the compiled Neuron models with checkpoint artiifacts (eg. config, tokenizer) will be stored by Optimum Neuron after the the first time compilation as well.

Reviewer: ```suggestion
 stores NEFF files in a local directory, usually `/var/tmp/neuron-compile-cache`. And for models adopting the second approach, the compiled Neuron models with checkpoint artifacts (eg. config, tokenizer) will be stored by Optimum Neuron after the the first time compilation as well.
```

>  stores NEFF files in a local directory, usually `/var/tmp/neuron-compile-cache`. And for models adopting the second approach, the compiled Neuron models with checkpoint artiifacts (eg. config, tokenizer) will be stored by Optimum Neuron after the the first time compilation as well.

Reviewer 2: ```suggestion
 stores NEFF files in a local directory, usually `/var/tmp/neuron-compile-cache`. And for models adopting the second approach, the compiled Neuron models with checkpoint artiifacts (eg. config, tokenizer) will be stored by `optimum-neuron` after the the first time compilation as well.
```

## How to use the Neuron model cache

The public model cache will be used when you use the `NeuronTrainer` or `NeuronModelForXXX` classes. There are no additional changes needed.

When exporting a model to neuron format, `optimum-neuron` will simply look for cached NEFF files during the compilation of the
model subgraphs or search for Neuron traced models in the hub repository .

If the NEFF files / Neuron models are cached, they will be fetched from the hub and directly loaded instead of being recompiled.

## How caching works

The Optimum Neuron Cache is built on top of the [NeuronX compiler cache](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/neuron-features/neuron-caching.html).

> It is important to understand that the cache of training / models applying `transformers_neuronx`operates on NEFF binaries, and not on the model itself.

Reviewer: ```suggestion
It is important to understand that the cache of training / models applying `transformers_neuronx` operates on NEFF binaries, and not on the model itself.
```

> It is important to understand that the cache of training / models applying `transformers_neuronx`operates on NEFF binaries, and not on the model itself.

Reviewer 2: I dont really understand this sentence. Especially the "training / models" part
```suggestion
It is important to understand that the cache of training / models applying `transformers_neuronx`operates on NEFF binaries, and not on the model itself.
```
